### ISSUE TITLE
instanceof checks erroneously error if firebase is included twice

### ISSUE THREAD

**Issue by appsforartists** [firebase/firebase-js-sdk#5216]
### Describe your environment

  * Operating System version: macOS 11.4
  * Browser version: Chrome 92.0.4515.107 (Official Build) (x86_64)
  * Firebase SDK version: 9.0.0-beta.7
  * Firebase Product: firestore

### Describe the problem

The `firebase` SDK uses `instanceof` to assert that `collection` receives a Firestore.  If two instances of `firebase` get included in the same bundle (e.g. because both the main package and the `yarn link`ed dependency have a `firebase` in their `node_modules`), this test will fail even if the two `firebase` imports are identical.

A bundle shouldn't include two copies of `firebase`, but `firebase` should also be resilient enough to not throw an error if this happens.

Bundling is hard.

#### Steps to reproduce:

1. Have two packages on your machine
2. Use `yarn link` to include `package-a` in `package-b`
3. Export the result of `initializeApp` from `package-a`
4. In `package-b`, call `collection` on a Firestore created from `package-a`'s export

#### Relevant Code:

```javascript
// package-a
export let firebaseApp: FirebaseApp = initializeApp(firebaseConfig);

// package-b
import {
  firebaseApp,
} from 'package-a';

const db = getFirestore(firebaseApp);

addDoc(
  collection(db, path),
  value
)
```

#### Error
```
Uncaught FirebaseError: Expected first argument to collection() to be a CollectionReference, a DocumentReference or FirebaseFirestore
```

where `firebaseApp` is an `Su extends _u` and `collection` expects a `_u$1`.

**Comment by schmidt-sebastian** [comment]
@appsforartists Thanks for reporting this.

I am a little afraid to change this, as this problem likely points to the fact that somehow two different Firestore versions made it into your build. If we supported types from an older version, we now have to maintain compatibility between these versions and any change to an internal API might constitute a breaking change to users that rely on this behavior. 
I would also like to say that from a size stand point, "pressuring" our users to only use one SDK version might actually be the right thing to do.

I do appreciate your input and see where you are coming from, but as stated, I am a bit afraid of the consequences this change would have and as such, I am going to close this for now.

**Comment by appsforartists** [comment]
As I see it:

- Duck typing is a better solution than `instanceof`, particularly in JavaScript.  Can you reasonably check the properties you need instead of the constructor type?
- Having two copies of Firebase in a bundle is indeed suboptimal.  It would be nice if the error message was better here.  Having to figure out WTF `Su` is and why its failing `t instanceof gu$1` is a frustrating timesink.

  Makes me wonder if the `instanceof` checks could do a duck check too and say "This looks like the Firestore from a different SDK. Only one instance of Firebase can be used at a time."
  
Admittedly, I had a hint that there were two SDKs in the bundle when `getApp()` failed.  Still, would have been nice to have a more explicit error.

---

Also, I appreciate the empathy in your reply.  In that vein, I offer these other thoughts, but I appreciate it was a weird bug to file in the first place. 🙂

**Comment by appsforartists** [comment]
Also worth noting, how I worked around this:

The library I was depending on was rolling up all its dependencies.  (It had been hosted on a CDN with previous versions of Firebase.)  That's how I ended up with two copies.

I used the `only` flag in the library's `rollup.config.js` to keep it from bundling its own copy of Firebase:

    resolve({
      // Make firebase external, so authors don't end up with duplicate copies
      // in their bundle
      only: [/^(?!firebase).*/],
    }),

and added a `dedupe` flag in my app's `rollup.config.js`:

    resolve({
      modulesOnly: true,
      dedupe: ['firebase'],
    }),

I might probably make a separate build for the CDN, that does its own bundling for consumers that don't have other uses for Firebase.

**Comment by schmidt-sebastian** [comment]
We have discussed using ducktyping to get around a similar problem before (which exists when users use `@google-cloud/firestore` together with `@firebase/firestore`). We decided against this since it usually only addresses the symptoms of a larger problem and might introduce unforeseen complications for our long term maintenance. I am glad to hear that you solved your problem and I hope that your advice can help other users.
